File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
    unordered_set<char>se;
  int leftPtr = 0 ,rightPtr = 0,n = s.length();
   if( n == 0) { return 0;}
   int ans = 0; 
     while(rightPtr < n){
        while(se.size() && se.find(s[rightPtr]) != se.end()){
             se.erase(s[leftPtr]);
            leftPtr++; 
        }
      se.insert(s[rightPtr]);
      ans = max(ans, rightPtr-leftPtr);
      rightPtr++;
     }
     return ans+1;
    }
};


/*


 genrate all substrings meanwhile check the  longest  substring without repeating
    tc : O(n*n*n)
    sc : O(26) :

  approach 2: 
  
    traverse using two ptrs meanshiwl stroing all the chars in set , 

     if char found : remove it from set  :  and also move leftptr till leftside reaching of char : 

      tc :O(n) : 
      sc : O(n)

       // "a b c b b c b b"

    //      a,b,c,b : 
    //      0,1,2,3 

"abcdefe" 
"abcbbcbb"
*/

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
      HashSet<Character>hs = new HashSet<>();
      int leftPtr = 0 , rightPtr = 0;
      int n = s.length();
      if(n ==0 ){return n;}
       int ans = 1;
      while(rightPtr < n) {
        char ch = s.charAt(rightPtr);
         while(leftPtr < rightPtr && hs.contains(ch)) {
            ans = Math.max(ans,rightPtr - leftPtr);
            hs.remove(s.charAt(leftPtr));
            leftPtr++;
         } 
        hs.add(ch); 
        rightPtr++;
      }
         ans = Math.max(ans,hs.size());
      return ans;  
    }
    /* 

    // au : missed TC:
      a b c a z e
       
  
    */
}